### PR TITLE
feat: GitLab preview deployments via webhook

### PR DESCRIPTION
**Pull request by edouard-mangel**
## Summary

- Adds preview deployment support for GitLab Merge Requests via webhooks, mirroring the existing GitHub preview deployment flow
- Exposes webhook secret configuration in the GitLab provider UI so users can register the webhook endpoint
- Implements permission-gated deployments: MR authors without write access are blocked and notified via an MR note

### Webhook handler (`pages/api/deploy/gitlab.ts`)
- Handles `Push Hook` (branch deploys with `triggerType = push`) and `Tag Push Hook` (tag deploys with `triggerType = tag`) as separate code paths — mirrors the GitHub handler pattern
- Handles `Merge Request Hook` actions: `open`, `update`, `reopen`, `labeled` → create/redeploy; `close`/`merge` → teardown preview deployments
- Permission check hoisted before the per-app loop: one API call per webhook event regardless of how many apps match
- Label-based filtering: only deploys if the MR carries a label configured on the application

### Server utilities (`packages/server/src/utils/providers/gitlab.ts`)
- `checkGitlabMemberPermissions`: verifies MR author has Developer+ access via GitLab Members API
- `hasExistingSecurityMRNote`: deduplicates security block notes (guards `response.ok` before JSON parse)
- `createSecurityBlockedMRNote` / `createMergeRequestNote`: posts/updates MR notes

### Services (`packages/server/src/services/gitlab.ts`)
- `findGitlabByWebhookSecret`: looks up the GitLab provider by webhook token (return type inferred, no `any`)

### Migration (`drizzle/0152_gitlab_preview_webhook.sql`)
- Adds nullable `webhook_secret` column, backfills with `encode(gen_random_bytes(21), 'base64')`, then sets `NOT NULL`

## Test plan

- [ ] Configure a GitLab provider with a webhook secret and register the webhook on a GitLab project pointing to `/api/deploy/gitlab`
- [ ] Open a Merge Request → preview deployment is created and queued
- [ ] Push to a branch → only apps with `triggerType = push` and matching branch are deployed
- [ ] Create a tag → only apps with `triggerType = tag` are deployed
- [ ] Open an MR as a user without Developer access → deployment is blocked and an MR note is posted
- [ ] Close/merge an MR → preview deployments are torn down

🤖 Generated with [Claude Code](https://claude.com/claude-code)

<h3>Greptile Summary</h3>

This PR implements GitLab preview deployment support via webhooks, mirroring the existing GitHub flow. It handles `Push Hook`, `Tag Push Hook`, and `Merge Request Hook` events, adds permission-gated deployments for MR authors, and exposes the webhook secret in the GitLab provider UI.

The implementation is generally well-structured and tests cover the key paths. Two issues were found:

- **P1 — Preview limit off-by-one** (`apps/dokploy/pages/api/deploy/gitlab.ts`): The guard `if (app.previewDeployments.length > previewLimit)` allows one extra deployment to be created when the count equals the limit exactly (e.g. 3 deployments with `previewLimit = 3` passes the check). The fix is to change `>` to `>=`.
- **P2 — Inaccurate `totalApps` count in Push Hook response**: The response message is derived from `apps.length + composeApps.length` before `watchPaths`-based filtering, so it can report \"Deployed N apps\" when fewer were actually queued.
- **P2 — Unhandled clipboard API rejection in `edit-gitlab-provider.tsx`**: `navigator.clipboard.writeText` returns a `Promise` that can reject in non-HTTPS or permission-denied contexts; the rejection is currently silently ignored and the success toast fires unconditionally.

<h3>Confidence Score: 4/5</h3>

Safe to merge after fixing the preview-limit off-by-one bug, which allows one more preview deployment than configured.

One P1 bug (preview limit off-by-one) means users can exceed their configured deployment limit by one. The rest of the implementation — permission checking, teardown logic, migration, and MR note deduplication — is correct and well-tested.

apps/dokploy/pages/api/deploy/gitlab.ts — preview-limit guard needs `>=` instead of `>`

<details><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| apps/dokploy/pages/api/deploy/gitlab.ts | New webhook handler for GitLab push, tag-push, and merge-request events. Contains a P1 off-by-one in the preview-limit guard (`>` should be `>=`), and a P2 inaccuracy in the `totalApps` response counter for push events when `watchPaths` filtering skips some apps. |
| packages/server/src/utils/providers/gitlab.ts | Adds GitLab member-permission checks and MR note utilities. Logic is sound: guards `response.ok` before JSON parse, deduplicates security notes via sentinel text, and correctly identifies Developer+ access (level >= 30). |
| packages/server/src/services/gitlab.ts | Adds `findGitlabByWebhookSecret` service function with clean return type inference (returns `null` when not found). Straightforward addition. |
| apps/dokploy/drizzle/0152_gitlab_preview_webhook.sql | Migration safely adds `webhook_secret` column: adds as nullable, backfills using `encode(gen_random_bytes(21), 'base64')`, then sets NOT NULL. |
| apps/dokploy/server/api/routers/gitlab.ts | Adds organization-scoping guard to the `one` query. Safe: `findGitlabById` already throws a `NOT_FOUND` TRPCError for unknown IDs. |
| apps/dokploy/components/dashboard/settings/git/gitlab/edit-gitlab-provider.tsx | Adds webhook configuration UI with copy buttons. The `copyToClipboard` helper doesn't handle Promise rejection from `navigator.clipboard.writeText`. |
| packages/server/src/db/schema/gitlab.ts | Adds `webhookSecret` column with `notNull()` constraint and `nanoid()` default. Matches the migration strategy for new rows. |

</details>

</details>

<details><summary><h3>Comments Outside Diff (1)</h3></summary>

1. `apps/dokploy/components/dashboard/settings/git/gitlab/edit-gitlab-provider.tsx`, line 826-834 ([link](https://github.com/dokploy/dokploy/blob/2035278f7d798faddfdedea9f2275d59ff1a748d/apps/dokploy/components/dashboard/settings/git/gitlab/edit-gitlab-provider.tsx#L826-L834)) 

   <a href="#">P2</a> **Unhandled rejection from `navigator.clipboard.writeText`**

   `navigator.clipboard.writeText` returns a `Promise` that rejects when clipboard access is unavailable (non-HTTPS context, permission denied, etc.). The rejection is silently swallowed here — `toast.success` fires and the copied state is set even before the write resolves. Consider awaiting the call inside an `async` handler with a `catch` that shows a toast error:

</details>

<sub>Reviews (1): Last reviewed commit: ["fix: address PR review issues in GitLab ..."](https://github.com/dokploy/dokploy/commit/2035278f7d798faddfdedea9f2275d59ff1a748d) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=26685565)</sub>

> Greptile also left **2 inline comments** on this PR.

<sub>(3/5) Reply to the agent's comments like "Can you suggest a fix for this @greptileai?" or ask follow-up questions!</sub>

### CODE REVIEW

File: apps/dokploy/pages/api/deploy/gitlab.ts [hunk 1]
```diff
@@ -0,0 +1,416 @@
...
+					const hasLabel = mrLabels.some((l) =>
+						app.previewLabels!.includes(l.title),
+					);
+					if (!hasLabel) continue;
+				}
+
+				// Preview limit
+				const previewLimit = app.previewLimit ?? 0;
+				if (app.previewDeployments.length > previewLimit) {
+					continue;
```
[the remark this answers is not in the document]
Comment (edouard-mangel) [reply apps/dokploy/pages/api/deploy/gitlab.ts:368]: Fixed in commit 6dcaafbf — changed `>` to `>=` so exactly N deployments at the configured limit correctly blocks the next one. Test updated to use the exact boundary case (3 existing == limit of 3) to verify the off-by-one is resolved.

File: apps/dokploy/pages/api/deploy/gitlab.ts [hunk 2]
```diff
@@ -0,0 +1,416 @@
...
+				}
+				await myQueue.add("deployments", { ...jobData }, {
+					removeOnComplete: true,
+					removeOnFail: true,
+				});
+			}
+
+			const totalApps = apps.length + composeApps.length;
+			res.status(200).json({
+				message:
```
[the remark this answers is not in the document]
Comment (edouard-mangel) [reply apps/dokploy/pages/api/deploy/gitlab.ts:115]: Fixed in commit 6dcaafbf — the Push Hook now tracks a \ variable incremented only when a job is actually dispatched (after the \ / \ filter passes). Note: this comment's line shifted to the Tag Push Hook block, which has no \ filter, so \ there remains accurate.

File: packages/server/src/utils/providers/gitlab.ts [hunk 3]
```diff
@@ -259,6 +259,199 @@ export const getGitlabBranches = async (input: {
...
+
+	const response = await fetch(
+		`${baseUrl}/api/v4/projects/${projectId}/merge_requests/${mrIid}/notes`,
+		{ headers: { Authorization: `Bearer ${gitlabProvider.accessToken}` } },
+	);
+	if (!response.ok) {
+		return false;
+	}
+	const notes: { id: number; body: string }[] = await response.json();
+	return notes.some((note) => note.body.includes(SECURITY_SENTINEL));
```
[the remark this answers is not in the document]
Comment (edouard-mangel) [reply packages/server/src/utils/providers/gitlab.ts]: Fixed in commit 6dcaafbf — replaced the single \ with a paginated loop that follows the \ header, requesting 100 notes per page until the sentinel is found or the header is empty. Tests updated to provide \ on mock responses.

File: apps/dokploy/pages/api/deploy/gitlab.ts [hunk 4]
```diff
@@ -0,0 +1,416 @@
...
+
+			if (blocked) {
+				await createSecurityBlockedMRNote(
+					gitlabProvider.gitlabId,
+					projectId,
+					mrIid,
+					mrAuthor,
+					pathNamespace,
+					blockedAccessLevel,
+				);
```
[the remark this answers is not in the document]
Comment (edouard-mangel) [reply apps/dokploy/pages/api/deploy/gitlab.ts]: Fixed in commit 6dcaafbf — \ is now wrapped in a \. A transient GitLab API error is logged but no longer throws, so authorized preview deployments are not blocked by a failed note post.

File: apps/dokploy/__test__/deploy/gitlab.test.ts [hunk 5]
```diff
@@ -0,0 +1,455 @@
...
+		expect(myQueue.add).not.toHaveBeenCalled();
+	});
+
+	it("reports security note with the blocked app's access level, not a later app's level", async () => {
+		// Two apps: first blocked (access_level=20), second allowed (access_level=30)
+		// The security note for the first app must report access_level=20, not 30
+		vi.mocked(checkGitlabMemberPermissions)
+			.mockResolvedValueOnce({ hasWriteAccess: false, accessLevel: 20 })
+			.mockResolvedValueOnce({ hasWriteAccess: true, accessLevel: 30 });
+
```
[the remark this answers is not in the document]
Comment (edouard-mangel) [reply apps/dokploy/__test__/deploy/gitlab.test.ts]: Fixed in commit 6dcaafbf — the test was updated to reflect the hoisted permission check. It now mocks \ with a single \ (not two \ calls), expects exactly 1 API call covering all apps, and asserts that all apps are blocked when access is denied (semantically correct: access is per-MR-author, not per-app).
